File: MapDemo/Objective C/MultipleShapesExample.m
//
//  MultipleShapesExample.m
//  MapDemo
//
//  Created by apple on 14/09/18.
//  Copyright © 2018 MMI. All rights reserved.
//

#import "MultipleShapesExample.h"
@import MapmyIndiaMaps;

@interface MultipleShapesExample () <MapmyIndiaMapViewDelegate>
@property (nonatomic) MapmyIndiaMapView *mapView;
@end

@implementation MultipleShapesExample

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.mapView = [[MapmyIndiaMapView alloc] initWithFrame:self.view.bounds];
    self.mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.mapView.tintColor = [UIColor darkGrayColor];
    
    [self.mapView setCenterCoordinate:CLLocationCoordinate2DMake(28.543253, 77.261647) zoomLevel:11 animated:NO];
    self.mapView.delegate = self;
    
    [self.view addSubview:self.mapView];
}

- (void)mapView:(MGLMapView *)mapView didFinishLoadingStyle:(MGLStyle *)style {
    
    // Parse the GeoJSON data.
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSString *path = [[NSBundle mainBundle] pathForResource:@"metro-line" ofType:@"geojson"];
        
        NSData *data = [NSData dataWithContentsOfFile:path];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self drawShapeCollection:data];
        });
    });
}

- (void)drawShapeCollection:(NSData *)data {
    
    // Use [MGLShape shapeWithData:encoding:error:] to create a MGLShapeCollectionFeature from GeoJSON data.
    MGLShape *feature = [MGLShape shapeWithData:data encoding:NSUTF8StringEncoding error:NULL];
    
    // Create source and add it to the map style.
    MGLShapeSource *source = [[MGLShapeSource alloc] initWithIdentifier:@"transit" shape:feature options:nil];
    [self.mapView.style addSource:source];
    
    // Create station style layer.
    MGLCircleStyleLayer *circleLayer = [[MGLCircleStyleLayer alloc] initWithIdentifier:@"stations" source:source];
    circleLayer.predicate = [NSPredicate predicateWithFormat:@"TYPE = 'Station'"];
    circleLayer.circleColor = [NSExpression expressionForConstantValue:[UIColor redColor]];
    circleLayer.circleRadius = [NSExpression expressionForConstantValue:@6];
    circleLayer.circleStrokeWidth = [NSExpression expressionForConstantValue:@2];
    circleLayer.circleStrokeColor = [NSExpression expressionForConstantValue:[UIColor blackColor]];
    
    // Create line style layer.
    MGLLineStyleLayer *lineLayer = [[MGLLineStyleLayer alloc] initWithIdentifier:@"rail-line" source: source];
    lineLayer.predicate = [NSPredicate predicateWithFormat:@"TYPE = 'Rail line'"];
    lineLayer.lineColor = [NSExpression expressionForConstantValue:[UIColor redColor]];
    lineLayer.lineWidth = [NSExpression expressionForConstantValue:@5];
    
    MGLFillStyleLayer *polygonLayer = [[MGLFillStyleLayer alloc] initWithIdentifier:@"station-boundry" source:source];
    polygonLayer.predicate = [NSPredicate predicateWithFormat: @"TYPE = 'Rail Station'"];
    polygonLayer.fillColor = [NSExpression expressionForConstantValue: UIColor.redColor];
    polygonLayer.fillOpacity = [NSExpression expressionForConstantValue: @0.5];
    polygonLayer.fillOutlineColor = [NSExpression expressionForConstantValue: UIColor.blackColor];
    
    // Add style layers to the map view's style.
    [self.mapView.style addLayer:circleLayer];
    [self.mapView.style insertLayer:lineLayer belowLayer:circleLayer];
    [self.mapView.style insertLayer:polygonLayer belowLayer:circleLayer];
}

@end

File: MapDemo/Swift/MultipleShapesExample.swift
//
//  MultipleShapesExample.swift
//  MapDemo
//
//  Created by apple on 14/09/18.
//  Copyright © 2018 MMI. All rights reserved.
//

import Mapbox

@objc(MultipleShapesExample_Swift)
class MultipleShapesExample_Swift: UIViewController, MGLMapViewDelegate {
    
    var mapView: MGLMapView!
     var strType:String?
    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = strType
        mapView = MapmyIndiaMapView(frame: view.bounds)        
        mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mapView.tintColor = .darkGray
        
        mapView.setCenter(CLLocationCoordinate2D(latitude: 28.543253, longitude: 77.261647), zoomLevel: 11, animated: false)
        
        mapView.delegate = self
        view.addSubview(mapView)
    }
    
    func mapView(_ mapView: MGLMapView, didFinishLoading style: MGLStyle) {
        
        // Parse the GeoJSON data.
        DispatchQueue.global().async {
            guard let url = Bundle.main.url(forResource: "metro-line", withExtension: "geojson") else {
                preconditionFailure("Failed to load local GeoJSON file")
            }
            
            guard let data = try? Data(contentsOf: url) else {
                preconditionFailure("Failed to decode GeoJSON file")
            }
            
            DispatchQueue.main.async {
                try? self.drawShapeCollection(data: data)
            }
        }
    }
    
    func drawShapeCollection(data: Data) throws {
        guard let style = self.mapView.style else { return }
        
        // Use [MGLShape shapeWithData:encoding:error:] to create a MGLShapeCollectionFeature from GeoJSON data.
        guard let feature = try? MGLShape(data: data, encoding: String.Encoding.utf8.rawValue) as? MGLShapeCollectionFeature else {
            fatalError("Could not cast to specified MGLShapeCollectionFeature")
        }
        
        // Create source and add it to the map style.
        let source = MGLShapeSource(identifier: "transit", shape: feature, options: nil)
        style.addSource(source)
        
        // Create station style layer.
        let circleLayer = MGLCircleStyleLayer(identifier: "stations", source: source)
        
        // Use a predicate to filter out non-points.
        circleLayer.predicate = NSPredicate(format: "TYPE = 'Station'")
        circleLayer.circleColor = NSExpression(forConstantValue: UIColor.red)
        circleLayer.circleRadius = NSExpression(forConstantValue: 6)
        circleLayer.circleStrokeWidth = NSExpression(forConstantValue: 2)
        circleLayer.circleStrokeColor = NSExpression(forConstantValue: UIColor.black)
        
        // Create line style layer.
        let lineLayer = MGLLineStyleLayer(identifier: "rail-line", source: source)
        
        // Use a predicate to filter out the stations.
        lineLayer.predicate = NSPredicate(format: "TYPE = 'Rail line'")
        lineLayer.lineColor = NSExpression(forConstantValue: UIColor.red)
        lineLayer.lineWidth = NSExpression(forConstantValue: 5)
        
        
        
        let polygonLayer = MGLFillStyleLayer(identifier: "station-boundry", source: source)
        polygonLayer.predicate = NSPredicate(format: "TYPE = 'Rail Station'")
        polygonLayer.fillColor = NSExpression(forConstantValue: UIColor.red)
        polygonLayer.fillOpacity = NSExpression(forConstantValue: 0.5)
        polygonLayer.fillOutlineColor = NSExpression(forConstantValue: UIColor.black)
        
        // Add style layers to the map view's style.
        style.addLayer(circleLayer)
        style.insertLayer(lineLayer, below: circleLayer)
        style.insertLayer(polygonLayer, below: circleLayer)
    }
}
